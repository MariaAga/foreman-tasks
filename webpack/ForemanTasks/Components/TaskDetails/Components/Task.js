import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid, Row, Col, Button } from 'patternfly-react';
import { translate as __ } from 'foremanReact/common/I18n';
import TaskInfo from './TaskInfo';
import { ClickConfirmation } from '../../common/ClickConfirmation';
import { ResumeButton } from '../../common/ActionButtons/ResumeButton';
import { CancelButton } from '../../common/ActionButtons/CancelButton';

class Task extends Component {
  taskProgressToggle = () => {
    const {
      timeoutId,
      refetchTaskDetails,
      id,
      loading,
      taskReloadStop,
      taskReloadStart,
    } = this.props;
    if (timeoutId) {
      taskReloadStop(timeoutId);
    } else {
      taskReloadStart(timeoutId, refetchTaskDetails, id, loading);
    }
  };

  render() {
    const {
      taskReload,
      externalId,
      id,
      state,
      allowDangerousActions,
      resumable,
      cancellable,
      hasSubTasks,
      parentTask,
      showUnlockModal,
      showForceUnlockModal,
      toggleUnlockModal,
      toggleForceUnlockModal,
      cancelTaskRequest,
      resumeTaskRequest,
      action,
      dynflowEnableConsole,
    } = this.props;
    const modalUnlock = (
      <ClickConfirmation
        showModal={showUnlockModal}
        title={__('Unlock')}
        body={__(
          "This will unlock the resources that the task is running against. Please note that this might lead to inconsistent state and should be used with caution, after making sure that the task can't be resumed."
        )}
        confirmationMessage={__(
          'I understand that this may cause harm and have working database backups of all backend services.'
        )}
        confirmAction={__('Unlock')}
        path={`/foreman_tasks/tasks/${id}/unlock`}
        confirmType="warning"
        closeModal={toggleUnlockModal}
      />
    );

    const modalForceUnlock = (
      <ClickConfirmation
        showModal={showForceUnlockModal}
        title={__('Force Unlock')}
        body={__(
          'Resources will be unlocked and will not prevent other tasks from being run. As the task might be still running, it should be avoided to use this unless you are really sure the task got stuck'
        )}
        confirmationMessage={__(
          'I understand that this may cause harm and have working database backups of all backend services.'
        )}
        confirmAction={__('Force Unlock')}
        path={`/foreman_tasks/tasks/${id}/force_unlock`}
        confirmType="danger"
        closeModal={toggleForceUnlockModal}
      />
    );
    return (
      <React.Fragment>
        {modalUnlock}
        {modalForceUnlock}
        <Grid>
          <Row>
            <Col xs={12}>
              <Button
                hidden={!allowDangerousActions}
                className="reload-button"
                bsSize="small"
                onClick={this.taskProgressToggle}
              >
                <span
                  className={`glyphicon glyphicon-refresh ${
                    taskReload ? 'spin' : ''
                  }`}
                />
                {__(`${taskReload ? 'Stop' : 'Start'}  auto-reloading`)}
              </Button>
              <Button
                bsSize="small"
                href={`/foreman_tasks/dynflow/${externalId}`}
                disabled={!dynflowEnableConsole}
              >
                {__('Dynflow console')}
              </Button>
              <ResumeButton
                id={id}
                onClick={() => {
                  if (!taskReload) {
                    this.taskProgressToggle();
                  }
                  resumeTaskRequest(id, action);
                }}
                name={action}
                disabled={!resumable}
              />
              <CancelButton
                id={id}
                name={action}
                disabled={!cancellable}
                onClick={() => {
                  if (!taskReload) {
                    this.taskProgressToggle();
                  }
                  cancelTaskRequest(id, action);
                }}
              />
              {parentTask && (
                <Button
                  bsSize="small"
                  href={`/foreman_tasks/tasks/${parentTask}`}
                >
                  {__('Parent task')}
                </Button>
              )}
              {hasSubTasks && (
                <Button
                  bsSize="small"
                  href={`/foreman_tasks/tasks/${id}/sub_tasks`}
                >
                  {__('Sub tasks')}
                </Button>
              )}
              {allowDangerousActions && (
                <Button
                  bsSize="small"
                  disabled={state !== 'paused'}
                  onClick={toggleUnlockModal}
                >
                  {__('Unlock')}
                </Button>
              )}
              {allowDangerousActions && (
                <Button
                  bsSize="small"
                  disabled={state === 'stopped'}
                  onClick={toggleForceUnlockModal}
                >
                  {__('Force Unlock')}
                </Button>
              )}
            </Col>
          </Row>
          <TaskInfo {...this.props} />
        </Grid>
      </React.Fragment>
    );
  }
}

Task.propTypes = {
  ...TaskInfo.PropTypes,
  state: PropTypes.string,
  allowDangerousActions: PropTypes.bool,
  resumable: PropTypes.bool,
  cancellable: PropTypes.bool,
  refetchTaskDetails: PropTypes.func,
  hasSubTasks: PropTypes.bool,
  parentTask: PropTypes.string,
  taskReload: PropTypes.bool,
  taskReloadStop: PropTypes.func,
  taskReloadStart: PropTypes.func,
  timeoutId: PropTypes.number,
  externalId: PropTypes.string,
  id: PropTypes.string.isRequired,
  showUnlockModal: PropTypes.bool,
  showForceUnlockModal: PropTypes.bool,
  toggleUnlockModal: PropTypes.func,
  toggleForceUnlockModal: PropTypes.func,
  cancelTaskRequest: PropTypes.func,
  resumeTaskRequest: PropTypes.func,
  dynflowEnableConsole: PropTypes.bool,
};

Task.defaultProps = {
  ...TaskInfo.defaultProps,
  state: '',
  allowDangerousActions: false,
  resumable: false,
  cancellable: false,
  refetchTaskDetails: () => null,
  hasSubTasks: false,
  parentTask: '',
  taskReload: false,
  taskReloadStop: () => null,
  taskReloadStart: () => null,
  timeoutId: null,
  externalId: '',
  showUnlockModal: false,
  showForceUnlockModal: false,
  toggleUnlockModal: () => null,
  toggleForceUnlockModal: () => null,
  cancelTaskRequest: () => null,
  resumeTaskRequest: () => null,
  dynflowEnableConsole: false,
};

export default Task;
