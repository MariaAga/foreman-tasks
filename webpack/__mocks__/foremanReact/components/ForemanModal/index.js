import React from 'react';

const ForemanModal = () => <div>Modal</div>;
ForemanModal.Header = () => <div>Modal Header</div>;
ForemanModal.Footer = () => <div>Modal Footer</div>;
export default ForemanModal;
