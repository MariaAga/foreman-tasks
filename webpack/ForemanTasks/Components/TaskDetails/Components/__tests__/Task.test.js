import React from 'react';
import {
  testComponentSnapshotsWithFixtures,
  mount,
  shallow,
} from '@theforeman/test';
import { useForemanModal } from 'foremanReact/components/ForemanModal/ForemanModalHooks';
import Task from '../Task';
import {
  UNLOCK_MODAL,
  FORCE_UNLOCK_MODAL,
} from '../../../TaskActions/TaskActionsConstants';

const fixtures = {
  'render without Props': { id: 'test' },
  'render with some Props': {
    id: 'test',
    state: 'paused',
    hasSubTasks: true,
    dynflowEnableConsole: true,
    parentTask: 'parent-id',
    taskReload: true,
    canEdit: true,
  },
};

describe('Task', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(Task, fixtures));
  describe('click test', () => {
    const setModalOpen = jest.fn();
    useForemanModal.mockImplementation(id => ({
      setModalOpen: () => setModalOpen(id),
    }));
    const cancelTaskRequest = jest.fn();
    const resumeTaskRequest = jest.fn();
    const taskProgressToggle = jest.fn();
    const id = 'some-id';
    const action = 'some-action';
    const props = {
      taskReload: false,
      id,
      action,
      cancelTaskRequest,
      resumeTaskRequest,
      taskProgressToggle,
    };
    afterEach(() => {
      jest.clearAllMocks();
    });
    it('reload', () => {
      const component = mount(<Task {...props} />);
      const reloadButton = component.find('.reload-button').at(0);
      reloadButton.simulate('click');
      expect(taskProgressToggle).toBeCalled();
    });
    it('resume', () => {
      const component = shallow(<Task {...props} />);
      const resumeButton = component.find('.resume-button').at(0);
      resumeButton.props().onClick();
      expect(taskProgressToggle).toBeCalled();
      expect(resumeTaskRequest).toBeCalledWith(id, action);
    });
    it('cancel', () => {
      const component = shallow(<Task {...props} />);
      const cancelButton = component.find('.cancel-button').at(0);
      cancelButton.props().onClick();
      expect(taskProgressToggle).toBeCalled();
      expect(cancelTaskRequest).toBeCalledWith(id, action);
    });
    it('unlock', () => {
      const component = shallow(<Task {...props} />);
      const unlockButton = component.find('.unlock-button').at(0);
      unlockButton.props().onClick();
      expect(setModalOpen).toBeCalledWith({ id: UNLOCK_MODAL });
    });
    it('focrce unlock', () => {
      const component = shallow(<Task {...props} />);
      const forceUnlockButton = component.find('.force-unlock-button').at(0);
      forceUnlockButton.props().onClick();
      expect(setModalOpen).toBeCalledWith({ id: FORCE_UNLOCK_MODAL });
    });
  });
});
