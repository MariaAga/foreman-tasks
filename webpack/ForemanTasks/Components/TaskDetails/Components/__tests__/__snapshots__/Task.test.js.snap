// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task rendering render with some Props 1`] = `
<Fragment>
  <UnlockModal
    id="unlockModal"
    onClick={[Function]}
  />
  <ForceUnlockModal
    id="forceUnlockModal"
    onClick={[Function]}
    selectedRowsLen={1}
  />
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        xs={12}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="reload-button"
          disabled={false}
          hidden={false}
          onClick={[Function]}
        >
          <span
            className="glyphicon glyphicon-refresh spin"
          />
          Stop  auto-reloading
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="dynflow-button"
          disabled={false}
          href="/foreman_tasks/dynflow/"
        >
          Dynflow console
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="resume-button"
          disabled={true}
          onClick={[Function]}
        >
          Resume
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="cancel-button"
          disabled={true}
          onClick={[Function]}
        >
          Cancel
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="parent-button"
          disabled={false}
          href="/foreman_tasks/tasks/parent-id"
        >
          Parent task
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="subtask-button"
          disabled={false}
          href="/foreman_tasks/tasks/test/sub_tasks"
        >
          Sub tasks
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="unlock-button"
          disabled={false}
          onClick={[MockFunction]}
        >
          Unlock
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="force-unlock-button"
          disabled={false}
          onClick={[MockFunction]}
        >
          Force Unlock
        </Button>
      </Col>
    </Row>
    <TaskInfo
      action=""
      allowDangerousActions={true}
      cancelTaskRequest={[Function]}
      cancellable={false}
      dynflowEnableConsole={true}
      endedAt=""
      error={Array []}
      externalId=""
      hasSubTasks={true}
      help=""
      id="test"
      output=""
      parentTask="parent-id"
      progress={0}
      refetchTaskDetails={[Function]}
      result="error"
      resumable={false}
      resumeTaskRequest={[Function]}
      startAt=""
      startBefore=""
      startedAt=""
      state="paused"
      taskReload={true}
      taskReloadStart={[Function]}
      taskReloadStop={[Function]}
      timeoutId={null}
      username=""
      usernamePath=""
    />
  </Grid>
</Fragment>
`;

exports[`Task rendering render without Props 1`] = `
<Fragment>
  <UnlockModal
    id="unlockModal"
    onClick={[Function]}
  />
  <ForceUnlockModal
    id="forceUnlockModal"
    onClick={[Function]}
    selectedRowsLen={1}
  />
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        xs={12}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="reload-button"
          disabled={false}
          hidden={true}
          onClick={[Function]}
        >
          <span
            className="glyphicon glyphicon-refresh "
          />
          Start  auto-reloading
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="dynflow-button"
          disabled={true}
          href="/foreman_tasks/dynflow/"
        >
          Dynflow console
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="resume-button"
          disabled={true}
          onClick={[Function]}
        >
          Resume
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="small"
          bsStyle="default"
          className="cancel-button"
          disabled={true}
          onClick={[Function]}
        >
          Cancel
        </Button>
      </Col>
    </Row>
    <TaskInfo
      action=""
      allowDangerousActions={false}
      cancelTaskRequest={[Function]}
      cancellable={false}
      dynflowEnableConsole={false}
      endedAt=""
      error={Array []}
      externalId=""
      hasSubTasks={false}
      help=""
      id="test"
      output=""
      parentTask=""
      progress={0}
      refetchTaskDetails={[Function]}
      result="error"
      resumable={false}
      resumeTaskRequest={[Function]}
      startAt=""
      startBefore=""
      startedAt=""
      state=""
      taskReload={false}
      taskReloadStart={[Function]}
      taskReloadStop={[Function]}
      timeoutId={null}
      username=""
      usernamePath=""
    />
  </Grid>
</Fragment>
`;
